### ISSUE TITLE
Spec should forbid annotations anywhere within `instanceof` types

### ISSUE THREAD

**Issue by cpovirk** [jspecify/jspecify#626]
(assigning to @netdpb in case he wants to include this in https://github.com/jspecify/jspecify/pull/592)

javac accepts code like this:

```java
import org.jspecify.annotations.Nullable;

class InstanceOfAnnotated {
  void go(Object o) {
    if (o instanceof Foo<? extends @Nullable Number>) {}
  }

  interface Foo<T extends Number> {}
}
```

We already ban annotations on the _root_ type of the `instanceof` types. We should also ban them on any subcomponent types:

> [INSTANCEOF -- we should not permit anything that a user could think was being checked but wasn't.](https://github.com/jspecify/jspecify/issues/17#issuecomment-484997275)

This will be a way in which casts differ from `instanceof`.

**Comment by agentgt** [comment]
Another oddity is that you can do this:

```java
String a = switch(o) {
  case @Nullable String l -> "stuff";
  case null -> "";
};
```

Which is fine for `javac` (as it has no idea what `@Nullable` is) but does not make sense with JSpecify.

I'm not sure if that is covered anywhere else.

**Comment by cpovirk** [comment]
Thanks, I think we had kind of resigned ourselves to sticking with the language features of Java 11ish, but we should at minimum track that for later. My hope would be that `case` patterns are similar to `instanceof` patterns. Does that seem believable?

Hmm, and that gets me to think about them in more detail. I guess that it's a little sad that we would be forbidding:

```java
Collection<? extends @Nullable String> collection = ...;
if (collection instanceof List<? extends @Nullable String> list) { ... }
```

It's especially sad that that's actually _worse_ than omitting the annotation, since omitting the annotation makes the type more specific, and the specificity is exactly what we're trying to communicate is not checked:

```java
Collection<? extends @Nullable String> collection = ...;
if (collection instanceof List<? extends String> list) { ... } // worse!
```

Maybe we "need" to allow the annotations (and then make checkers responsible for making sure that they match?)? Otherwise, users have to use plain `instanceof List`, which can't be used with a pattern-style variable declaration (unless you want the raw type, yuck), IIRC.

So this needs more thought, and it is not a good candidate for sneaking into #592.

(Well, we could still consider forbidding them in spec 1.0 and then being more permissive later.)

**Comment by agentgt** [comment]
For sealed classes I think we forbid doing something like this right?

```java
sealed interface Parent<T extends @Nullable Object> {
  record SomeThing<T extends @NonNull Object>() implements Parent<T> {
  }
}
```

**Comment by cpovirk** [comment]
Hmm, I would think we'd probably allow that. That does sound like a potential further complication, in that it would make sense to allow `case SomeThing<? extends @NonNull Object>` when switching on a `Parent<? extends @Nullable Object>`. Or maybe that would be covered naturally by the vaporware rule that we should already have? The good news, at least, is that `case SomeThing<?>` should work fine there, since it's essentially equivalent to `case SomeThing<? extends @NonNull Object>`, given the non-nullable type-parameter declaration.

**Comment by agentgt** [comment]
Yes I agree but at the moment the spec seems to make that the above code is not allowed or am I wrong?

**Comment by cpovirk** [comment]
Yeah, I was careful in my phrasing above that "javac accepts code," as I suspect that you're right that javac is doing things that aren't actually backed up by the spec :)

We may see `? extends Object` used even in spec-compliant code: The spec does have a point somewhere that `?` and `? extends Object` are equivalent, and I think there were some javac changes to avoid issuing warnings for some usages of `? extends Object` on those grounds. It's highly possible that those changes have not been fully plumbed through the spec and that they're otherwise not fully baked. (Certainly JSpecify depends on the difference between `?` and `? extends Object`. Fortunately, I haven't seen any further signs that we're going to get into trouble for that.)

**Comment by cpovirk** [comment]
> I think we had kind of resigned ourselves to sticking with the language features of Java 11ish

On that front, I note that even `instanceof` with a semi-unbounded type...

```java
if (o instanceof Foo<? extends Object>) {}
```

...produces an error if built with javac 11. Of course, we'll continue to nudge people to [build with a newer version of javac](https://github.com/jspecify/jspecify/wiki/version-compatibility#very-rough-notes-on-selecting-a-jdk-to-build-with) even when targeting older runtimes. But there is some argument for holding the line that language features in releases after Java 11 are out of scope for now. (And that may argue either for banning new possible annotation locations or for not speaking to them at all. Either way, we can revisit later.)

(Interestingly, javac 22 accepts the code above even with `-source 11` / `--release 11`. Not that there's any reason to write that code: The type arguments pay off only if you are declaring a variable, as in [my `List<? extends @Nullable String> list` example above](https://github.com/jspecify/jspecify/issues/626#issuecomment-2338535355). I just found that interesting, given that javac (IIRC) rejects at least one other compile-time-only feature, the diamond operator, when targeting older runtimes.)

**Comment by cpovirk** [comment]
Per https://github.com/jspecify/jspecify/pull/644, these will stay unrecognized for spec 1.0.0 on the theory that we don't want to block that release on a further discussion and we can always make the rule more lenient later. I'm removing this issue from the milestone but leaving the issue open for future discussion.

**Comment by cpovirk** [comment]
https://github.com/jspecify/jspecify/pull/668 makes clearer that there could be two different decisions here: one for plain old `instanceof Foo` and another for `instanceof Foo foo` (along with, presumably, the patterns in `switch`, too).

**Comment by cpovirk** [comment]
@wmdietl may have me [convinced](https://github.com/jspecify/jspecify/pull/679#pullrequestreview-2378887162) that we should eventually allow annotations in both cases.

**Comment by cpovirk** [comment]
I note that Java lets you _generalize_ the type argument as part of `instanceof`, whether you're doing just the "type comparison operator" or the "pattern":

```java
  boolean d(Iterable<String> i) {
    return i instanceof List<? extends CharSequence> list; // or without "list"
  }
```

Presumably nullness checkers would allow you to generalize the nullness of the type argument, too.

**Comment by netdpb** [comment]
In that example, why would I type `List<? extends CharSequence>` rather than `List<String>` there?

**Comment by cpovirk** [comment]
¯\\_(ツ)_/¯

**Comment by cpovirk** [comment]
Ah, maybe the idea is that you could sometimes get away with `List<?>` instead of `List<SomeComplicatedTypeGoesHere>`—or even `Map<Foo, ?>` instead of `Map<Foo, SomeComplicatedTypeGoesHere>`. So that gives them enough reason to _allow_ you to pick a more general type. And if you pick `List<? extends CharSequence>` rather than `List<String>`, then, well, that's not what they were going for :)

**Comment by netdpb** [comment]
Ah, but `List<?>` is covered by the _other_ way in which a narrowing conversion is checked: ["All of the type arguments of T are unbounded wildcards"](https://docs.oracle.com/javase/specs/jls/se23/html/jls-5.html#jls-5.1.6.2). Why allow something in between?

**Comment by cpovirk** [comment]
Does anything else cover the case of (say) `Map<Foo, SomeComplicatedTypeGoesHere>` -> `SortedMap<Foo, ?>` case? I'd expect that to be unusual but perhaps handy on occasion. (too lazy to check myself, since you two appear to already be on the case)

**Comment by cpovirk** [comment]
I'm noticing (in CL 739291162, for any Googlers) that it would be convenient to be able to write:

```java
Collection<E> collection = ...;
if (collection instanceof ImmutableCollection<@NonNull E> immutable) { ... }
```

I've been talking above as if that would need to be a nullness error, since `E` and `@NonNull E` don't "match." But given that `ImmutableCollection` requires a non-null type argument, it is actually safe. Maybe we will want to investigate a way to allow it. It might be somewhat analogous to allowing a `Collection<?> c` to be tested with `c instanceof EnumSet<? extends Enum<?>> enumSet`, as javac allows. (I'd have to look at the JLS for further information.)

Really, this is more general than `instanceof`/patterns: It applies to casts, as well, for which I also think we're currently lacking rules.
